src/ex03_project/test_sorting.py

```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+
+__author__ = 'Kaspar Akilles Lilja'
+__email__ = 'kalilja@nmbu.no'
+
+
+def bubble_sort(data):
+    datalist = list(data)
+    for i in range(0, len(datalist) - 1):
+        for j in range(0, len(datalist) - 1 - i):
+            if datalist[j] > datalist[j + 1]:
+                datalist[j], datalist[j + 1] = datalist[j + 1], datalist[j]
+    return datalist
+
+
+def test_empty():
+    """Test that the sorting function works for empty list"""
+    empty_list = []
+    assert bubble_sort(empty_list) == empty_list
+
+
+def test_single():
+    """Test that the sorting function works for single-element list"""
+    single_element_list = [2]
+    assert bubble_sort(single_element_list) == single_element_list
+
+
+def test_sorted_is_not_original():
+    """
+    Test that the sorting function returns a new object.
+    """
+    data = [3, 2, 1]
+    sorted_data = bubble_sort(data)
+    assert id(data) != id(sorted_data)
+
+
+def test_original_unchanged():
+    """
+    Test that sorting leaves the original data unchanged.
+    """
+    data = [3, 2, 1]
+    sorted_data = bubble_sort(data)
+    assert data == [3, 2, 1] and sorted_data != [3, 2, 1]
+
+
+def test_sort_sorted():
+    """Test that sorting works on sorted data."""
+    data = [1, 2, 3, 4, 6, 9]
+    sorted_data = bubble_sort(data)
+    for small, large in zip(sorted_data[:-1], sorted_data[1:]):
```

Comment:
Cool! `zip` will always return something with the same length as the shortest of its arguments:
```
In [16]: list(zip([1, 2],[1, 2, 3, 4]))                                         
Out[16]: [(1, 1), (2, 2)]
```